src/java_bytecode/java_types.h

```diff
@@ -313,6 +319,64 @@ inline const typet &java_generic_class_type_bound(size_t index, const typet &t)
   return gen_type.subtype();
 }
 
+/// Type to hold a Java class that is implicitly generic, e.g., an inner
+/// class of a generic outer class or a derived class of a generic base
+/// class. Extends the java class type.
+class java_implicitly_generic_class_typet:public java_class_typet
+{
+public:
+  typedef std::vector<java_generic_parametert> implicit_generic_typest;
+
+  explicit java_implicitly_generic_class_typet(
+    const java_class_typet &class_type,
+    const implicit_generic_typest &generic_types)
+    : java_class_typet(class_type)
+  {
+    set(ID_C_java_implicitly_generic_class_type, true);
+    for(const auto &type : generic_types)
+    {
+      implicit_generic_types().push_back(type);
+    }
+  }
+
+  const implicit_generic_typest &implicit_generic_types() const
+  {
+    return (const implicit_generic_typest&)(
+      find(ID_implicit_generic_types).get_sub());
+  }
+
+  implicit_generic_typest &implicit_generic_types()
+  {
+    return (implicit_generic_typest &)(
+      add(ID_implicit_generic_types).get_sub());
+  }
+};
+
+/// \param type: the type to check
+/// \return true if type is a implicitly generic java class type
+inline bool is_java_implicitly_generic_class_type(const typet &type)
```

Comment:
I may be wrong, but you may also need a version of this function that accepts a non-`const` `typet`

Reply:
@smowton Could you advise on this, is another version needed?

Reply:
No, non-const is implicitly castable to const, and this is inherently a read-only operation. @NlightNFotis is probably thinking of the `to_*` cast functions, which do want a const and a non-const flavour.